Write out some set_ overloads instead of relying on code binding generation.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28597 Kill _th_zero binding, just use a simple native function instead.
* #28596 Kill THTensor_(match), which isn't used.
* **#28595 Write out some set_ overloads instead of relying on code binding generation.**

Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

gchanan committed:
Write out some set_ overloads instead of relying on code binding generation.

[ghstack-poisoned]

gchanan committed:
Update on "Write out some set_ overloads instead of relying on code binding generation."


Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

[ghstack-poisoned]

gchanan committed:
Update on "Write out some set_ overloads instead of relying on code binding generation."


Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

[ghstack-poisoned]

gchanan commented:
Failures look plausibly caused by this, although I haven't been able to reproduce them locally.

albanD commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -34,6 +34,15 @@ Tensor _shape_as_tensor(const Tensor& self) {
 }
 
+Tensor& set_(Tensor& result) {
+  Storage storage(result.dtype(), 0, c10::GetAllocator(result.device().type()), true);
```
Isn't the storage supposed to be a nullptr from the previous signature in case it is not given by the user?

gchanan replied:
yes, but we can't don't actually support passing nullptrs in this way through the ATen->TH binding layer.

So what this PR attempts to do instead is to create a Storage in the same way the TH code does if you pass in a nullptr (you can't have nullptr Storage, so TH does this logic already).

albanD replied:
Right, the fix looks like the right thing to do :)

gchanan committed:
Update on "Write out some set_ overloads instead of relying on code binding generation."


Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

[ghstack-poisoned]

gchanan committed:
Update on "Write out some set_ overloads instead of relying on code binding generation."


Differential Revision: [D18116720](https://our.internmc.facebook.com/intern/diff/D18116720)

[ghstack-poisoned]
